lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/SchemaResolver.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2015 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.avro.compiler.idl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.avro.JsonProperties;
+import org.apache.avro.LogicalType;
+import org.apache.avro.Protocol;
+import org.apache.avro.Schema;
+import org.codehaus.jackson.JsonNode;
+
+/**
+ * Utility class to resolve schemas that are unavailable at the time they are referenced in the IDL.
+ */
+
+final class SchemaResolver {
+
+  private SchemaResolver() {
+  }
+
+  private static final String UR_SCHEMA_ATTR = "org.apache.avro.compiler.idl.unresolved.name";
+
+  static Schema unresolvedSchema(final String name) {
+    Schema schema = Schema.createRecord("UnresolvedSchema", "unresolved schema",
+            "org.apache.avro.compiler", false, Collections.EMPTY_LIST);
+    schema.addProp(UR_SCHEMA_ATTR, name);
+    return schema;
+  }
+
+  static boolean isUnresolvedSchema(final Schema schema) {
+    return (schema.getType() == Schema.Type.RECORD && schema.getProp(UR_SCHEMA_ATTR) != null);
+  }
+
+  static String getUnresolvedSchemaName(final Schema schema) {
+    String name = schema.getProp(UR_SCHEMA_ATTR);
+    if (name == null) {
+      throw new IllegalArgumentException("Schema " + schema + " is not a unresolved schema");
+    } else {
+      return name;
+    }
+  }
+
+  static Protocol resolve(final Protocol protocol) {
+    Protocol result = new Protocol(protocol.getName(), protocol.getDoc(), protocol.getNamespace());
+    final Collection<Schema> types = protocol.getTypes();
+    List<Schema> newSchemas = new ArrayList(types.size());
+    Map<String, Schema> processed = new HashMap<String, Schema>();
+    for (Schema schema : types) {
+      newSchemas.add(resolve(schema, protocol, processed));
+    }
+    result.setTypes(newSchemas);
+
+    for (Map.Entry<String, Protocol.Message> entry : protocol.getMessages().entrySet()) {
+      Protocol.Message value = entry.getValue();
+      Protocol.Message nvalue;
+      if (value.isOneWay()) {
+        Schema request = value.getRequest();
+        nvalue = result.createMessage(value.getName(), value.getDoc(),
+                value.getObjectProps(), intern(request, processed));
```

Comment:
Since I create new schemas instead of mutating, this "interning" just makes sure that any schemas referenced by a message that have been "resolved" are replaced with the new instances I created previously.